### PR TITLE
autoReload after an event on an Extension

### PR DESCRIPTION
**Pull request by colas31**
In our company we have automatize the installation of several private extension, but due to the popin to ask to reload VSCode, the process is blocked. 
To avoid user interaction, we would like to let automatically reload VSCode when it's needed.

### CODE REVIEW

File: extension/src/install.ts [hunk 1]
```diff
@@ -89,18 +92,36 @@ export async function updateExtensions(extensionInfo: ExtensionInfoService, pack
...
 
 /**
  * Displays a message with a button to reload vscode.
- * @param message The message to display.
+ * @param suffixMessage The reason to reload VSCode. The message will be displayed in a "popin".
  */
-export async function showReloadPrompt(message: string): Promise<void> {
-    const reload = await vscode.window.showInformationMessage(message, localize('reload.now', 'Reload Now'));
+export async function showReloadPrompt(suffixMessage: string): Promise<void> {
+    let reload: any;
```
Comment (joelspadin-garmin) [inline extension/src/install.ts:106]:
```suggestion
    let reload: boolean;
```

File: extension/src/install.ts [hunk 2]
```diff
@@ -89,18 +92,36 @@ export async function updateExtensions(extensionInfo: ExtensionInfoService, pack
...
+                'to complete updating the extensions.',
             ),
         );
     }
 }
 
 /**
  * Displays a message with a button to reload vscode.
- * @param message The message to display.
+ * @param suffixMessage The reason to reload VSCode. The message will be displayed in a "popin".
```
Comment (joelspadin-garmin) [inline extension/src/install.ts:103]: This will work for English, but I don't know if it will work for every language. Here are a couple ideas I came up with:

Idea 1: undo the changes to this parameter and just use a generic message when auto reloading:

```ts
export async function showReloadPrompt(message: string): Promise<void> {
    if (getAutoReload()) {
        message = localize(
            'automatic.reload',
            'Visual Studio Code will restart in 3 seconds to complete changes to private extensions.'
        );
        // ...
    } else {
        reload = await vscode.window.showInformationMessage(message, localize('reload.now', 'Reload Now'));
    }
```

Idea 2: pass in enough information for showReloadPrompt() to build the entire message.
``` ts
enum ReloadReason {
    Install,
    Uninstall,
    Update,
    UpdateAll,
}

function getReloadPromptMessage(reason: ReloadReason, extension?: string) {
    switch (reason) {
        case ReloadReason.Install:
            return localize(
                'reload.to.complete.install',
                'Please reload Visual Studio Code to complete installing the extension {0}.',
                extension,
            );
        // ...
    }
}

function getAutoReloadMessage(reason: ReloadReason, extension?: string) {
    switch (reason) {
        case ReloadReason.Install:
            return localize(
                'reload.to.complete.install',
                'Visual Studio Code will restart in 3 seconds to complete installing the extension {0}.',
                extension,
            );
        // ...
    }
}

/**
 * Displays a message with a button to reload vscode.
 * @param message The message to display.
 * @param name of the extension that was changed.
 */
export async function showReloadPrompt(reason: ReloadReason, extension?: string): Promise<void> {
    if (getAutoReload()) {
        const message = getAutoReloadMessage(reason, extension);
        // ...
    } else {
        const message = getReloadPromptMessage(reason, extension);
        // ...
    }
```

File: extension/src/install.ts [hunk 3]
```diff
@@ -89,18 +92,36 @@ export async function updateExtensions(extensionInfo: ExtensionInfoService, pack
...
+export async function showReloadPrompt(suffixMessage: string): Promise<void> {
+    let reload: any;
+
+    if (getAutoReload()) {
+        const message = localize(
+            'automatic.reload',
+            'Visual Studio Code will restart in 3 secondes {0}.',
+            suffixMessage,
+        );
+        vscode.window.showInformationMessage(message);
```
Comment (joelspadin-garmin) [inline extension/src/install.ts:114]: VS Code is usually pretty good about preserving state across a reload, but I would still be concerned about potential data loss. What if we lengthened the timer and provided an option to cancel the reload?

I think something like this might work:
```ts
const CANCEL_MESSAGE = localize('cancel', 'Cancel');
// ...

    const cancel = vscode.window.showInformationMessage(message, CANCEL_MESSAGE);
    const timeout = sleep(10000);
    const result = await Promise.race([cancel, timeout]);
    reload = result === CANCEL_MESSAGE;
```

**Review by joelspadin-garmin — CHANGES_REQUESTED** [review]
I'm not sure I fully understand the use case for this. VS Code doesn't have an option to do this for regular extensions from the marketplace. It will silently update the extensions in the background, but you still need to reload for the updates to apply. Does that also cause problems for you? 

I think having this option enabled might be annoying when manually installing extensions, as it will reload VS Code after every install. Maybe the auto reload should only apply to https://github.com/joelspadin-garmin/vscode-private-extension-manager/blob/855995109519e2ee93c876c82855010030ff2ae8/extension/src/install.ts#L66 instead of everything that uses showReloadPrompt()?

**Comment by colas31** [comment]
> I'm not sure I fully understand the use case for this. VS Code doesn't have an option to do this for regular extensions from the marketplace. It will silently update the extensions in the background, but you still need to reload for the updates to apply. Does that also cause problems for you?
> 
> I think having this option enabled might be annoying when manually installing extensions, as it will reload VS Code after every install. Maybe the auto reload should only apply to
> 
> https://github.com/joelspadin-garmin/vscode-private-extension-manager/blob/855995109519e2ee93c876c82855010030ff2ae8/extension/src/install.ts#L66
> 
> instead of everything that uses showReloadPrompt()?

Let me rephrase it :) 

Internally I have developed a private extension to allow developer to configure quickly their VSCode environment. Depending the king of project they are working on, they can install different "packages". Each composed composed of several public and private extension. 

I iterate over all extension and install them automatically, for public extension VSCode just displayed a button "Reload required" next to the name to the extension but for private extension there are a popin displayed this popin block my full process. User has to click on them to unlock it.

It's why I wanted to add this feature. 
But as you suggested I can indeed implement it only for installation and update of extensions. 

By default this setting is set to false, so no impact until the user change the value.

File: extension/src/install.ts (hunk 3 above)
Comment (colas31) [reply extension/src/install.ts:114]: I run some test locally, when user doesn't have the autoSave, the change are not lost.
But I like you idea to provide a cancel option :)

File: extension/src/install.ts [hunk 4]
```diff
@@ -89,18 +92,36 @@ export async function updateExtensions(extensionInfo: ExtensionInfoService, pack
...
  */
-export async function showReloadPrompt(message: string): Promise<void> {
-    const reload = await vscode.window.showInformationMessage(message, localize('reload.now', 'Reload Now'));
+export async function showReloadPrompt(suffixMessage: string): Promise<void> {
+    let reload: any;
+
+    if (getAutoReload()) {
+        const message = localize(
+            'automatic.reload',
+            'Visual Studio Code will restart in 3 secondes {0}.',
```
[the remark this answers is not in the document]
Comment (colas31) [reply extension/src/install.ts:111]: Oups, french typo ;)

File: extension/src/install.ts (hunk 2 above)
Comment (colas31) [reply extension/src/install.ts]: I keep the second idea :)

**Comment by joelspadin-garmin** [comment]
Oh, is your extension sending a `privateExtensions.extension.install` message to this extension in order to cause it to install the extensions for you? That wasn't really a supported use case, but if that is how you are using it I think there is a less invasive way to do that:

You could change
https://github.com/joelspadin-garmin/vscode-private-extension-manager/blob/855995109519e2ee93c876c82855010030ff2ae8/extension/src/commands/extensionCommands.ts#L40

to take an additional, optional parameter that prevents showing dialogs, for example:

```ts
export interface ExtensionCommandOptions {
    silent?: boolean;
}

// ...

    public async execute(extensionOrId: Package | string, options?: ExtensionCommandOptions): Promise<void> {
        const pkg = await this.extensionInfo.waitForExtensionChange(
            installExtension(this.registryProvider, extensionOrId),
        );

        // If vscode could immediately load the extension, it should be visible
        // to the extensions API now. If not, we need to reload.
        if (!options?.silent && !(await this.extensionInfo.getExtension(pkg.extensionId))) {
            await showInstallReloadPrompt(pkg);
        }
    }

**Comment by colas31** [comment]
> Oh, is your extension sending a `privateExtensions.extension.install` message to this extension in order to cause it to install the extensions for you? That wasn't really a supported use case, but if that is how you are using it I think there is a less invasive way to do that:
> 
> You could change
> https://github.com/joelspadin-garmin/vscode-private-extension-manager/blob/855995109519e2ee93c876c82855010030ff2ae8/extension/src/commands/extensionCommands.ts#L40
> 
> to take an additional, optional parameter that prevents showing dialogs, for example:
> 
> ```ts
> export interface ExtensionCommandOptions {
>     silent?: boolean;
> }
> 
> // ...
> 
>     public async execute(extensionOrId: Package | string, options?: ExtensionCommandOptions): Promise<void> {
>         const pkg = await this.extensionInfo.waitForExtensionChange(
>             installExtension(this.registryProvider, extensionOrId),
>         );
> 
>         // If vscode could immediately load the extension, it should be visible
>         // to the extensions API now. If not, we need to reload.
>         if (!options?.silent && !(await this.extensionInfo.getExtension(pkg.extensionId))) {
>             await showInstallReloadPrompt(pkg);
>         }
>     }
> ```

This approach is much better than what I have in mind. 

Thanks very much for your time. I have created this new PR instead https://github.com/joelspadin-garmin/vscode-private-extension-manager/pull/61
